File: Crawler/Result/Storage/MemoryStorage.php
<?php

namespace Crawler\Result\Storage;

use Crawler\Result\Storage\Storage;

class MemoryStorage implements Storage
{
    private array $storage = [];
    private bool $compress;

    /**
     * @param bool $compress
     */
    public function __construct(bool $compress = false)
    {
        $this->compress = $compress;
    }


    public function save(string $uqId, string $content): void
    {
        if ($this->compress) {
            $content = gzencode($content);
        }
        $this->storage[$uqId] = $content;
    }

    public function load(string $uqId): string
    {
        $content = $this->storage[$uqId] ?? '';
        if ($content && $this->compress) {
            $content = gzdecode($content);
        }
        return $content;
    }

    public function delete(string $uqId): void
    {
        unset($this->storage[$uqId]);
    }

    public function deleteAll(): void
    {
        $this->storage = [];
    }
}

File: src/result/storage/memory_storage.rs
// SiteOne Crawler - MemoryStorage
// (c) Jan Reges <jan.reges@siteone.cz>

use std::collections::HashMap;
use std::io::Write;

use crate::error::{CrawlerError, CrawlerResult};
use crate::result::storage::storage::Storage;

pub struct MemoryStorage {
    storage: HashMap<String, Vec<u8>>,
    compress: bool,
}

impl MemoryStorage {
    pub fn new(compress: bool) -> Self {
        Self {
            storage: HashMap::new(),
            compress,
        }
    }
}

impl Storage for MemoryStorage {
    fn save(&mut self, uq_id: &str, content: &[u8]) -> CrawlerResult<()> {
        let data = if self.compress {
            let mut encoder = flate2::write::GzEncoder::new(Vec::new(), flate2::Compression::default());
            encoder.write_all(content).map_err(CrawlerError::Io)?;
            encoder.finish().map_err(CrawlerError::Io)?
        } else {
            content.to_vec()
        };

        self.storage.insert(uq_id.to_string(), data);
        Ok(())
    }

    fn load(&self, uq_id: &str) -> CrawlerResult<Vec<u8>> {
        match self.storage.get(uq_id) {
            Some(data) if !data.is_empty() => {
                if self.compress {
                    let mut decoder = flate2::read::GzDecoder::new(&data[..]);
                    let mut decompressed = Vec::new();
                    std::io::Read::read_to_end(&mut decoder, &mut decompressed).map_err(CrawlerError::Io)?;
                    Ok(decompressed)
                } else {
                    Ok(data.clone())
                }
            }
            _ => Ok(Vec::new()),
        }
    }

    fn delete(&mut self, uq_id: &str) -> CrawlerResult<()> {
        self.storage.remove(uq_id);
        Ok(())
    }

    fn delete_all(&mut self) -> CrawlerResult<()> {
        self.storage.clear();
        Ok(())
    }
}

impl std::fmt::Debug for MemoryStorage {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("MemoryStorage")
            .field("entries", &self.storage.len())
            .field("compress", &self.compress)
            .finish()
    }
}
